### ISSUE TITLE
[css-sizing] Removing intrinsic aspect-ratio from an image

### ISSUE THREAD

**Issue by mirisuzanne** [w3c/csswg-drafts#6257]
The [current specification for `aspect-ratio`](https://drafts.csswg.org/css-sizing-4/#aspect-ratio) allows explicit ratios, or the "natural" aspect ratio from a replaced element. But there is not currently any explicit way to _remove_ the natural aspect ratio from replaced elements.

In grid & flexbox context, it is often useful to remove the natural ratio of an image, apply `object-*` properties, and allow the images to stretch in whatever space is available or defined by the grid-template. Currently, I'm not sure there's a way to do it without adding a wrapper, and [using absolute position](https://codepen.io/mirisuzanne/pen/poRqXGd).

It seems like we should be able to do this with something like `aspect-ratio: none`?

**Comment by tabatkins** [comment]
That's pretty reasonable, yeah. `none` definitely makes sense as the value for it.

**Comment by bfgeek** [comment]
We already have a mechanism to do this with `contain: size` which will make replaced elements have no aspect-ratio, and no natural size. The issue with `aspect-ratio: none` is that it wouldn't necessarily clear out intrinsic-size which might not be what people want.

**Comment by aethanyc** [comment]
https://github.com/w3c/csswg-drafts/issues/5550

**Comment by tabatkins** [comment]
@bfgeek Your question led me to suddenly realize I'm not sure what would happen to the natural sizes of a 300x150 image with `aspect-ratio: 1/1`, let alone `none`.

**Comment by Loirooriol** [comment]
Since `object-fit` doesn't seem to work well with `contain: size`, another approach can be:

```css
width: 0;
height: 0;
min-width: 100%;
min-width: 100%;
```

This should avoid natural sizes, 300x150 default intrinsic size, etc.

**Comment by bfgeek** [comment]
Looking at the current behaviour of `object-fit` with `contain: size` makes me pretty convinced that it's broken today.

**Comment by bfgeek** [comment]
Specifically - `object-fit` algorithm is using the "natural size" which `contain: size` is zeroing out. A potential improvement would be for `contain: size` to have an affect on the "natural size" for layout purposes only.

**Comment by tabatkins** [comment]
Yeah, that interaction does seem to be implied by the spec today, so good on impls for following the spec, but I agree it's a *broken* behavior and we should change it, probably in exactly the way you suggest.

**Comment by Loirooriol** [comment]
It's true that https://drafts.csswg.org/css-contain/#sizing-as-if-empty says

> Replaced elements must be treated as having an natural width and height of 0 and no natural aspect ratio.

But this is just for the purpose of sizing the `<img>` element or whatever. https://drafts.csswg.org/css-contain/#laying-out-in-place says

> The containment box's content (including any pseudo-elements) must then be laid out into the now fixed-size containment box normally.

So to me it doesn't seem like implementations are following that?

**Comment by jensimmons** [comment]
I won't be able to attend the 7pm CSSWG call today — it's too late in the day for me to have the energy. But this seems cool. 

To better understand what's currently possible, I wrote this demo: https://codepen.io/jensimmons/pen/3c9e5163fb70b7fb25adad0db97494fa?editors=1100
Aspect ratios (natural or set as preferred) only works if one of the dimensions is `auto`, so an alternative solution in some usecases is to set both `width` and `height` to something besides `auto`, like `100%`.  

Even if other piles of code can reach the same result, `aspect-ratio: none;` is a easily understandable straightforward solution. And elegance is good.

I'd love to see what @mirisuzanne was coding up. 

I wouldn't want this to only work for images / `<img>`, but instead for any & all elements that have a natural aspect ratio — including `<video>`.

**Comment by Loirooriol** [comment]
@jensimmons Note, though, that only works because the row has a fixed height. If you change `grid-template-rows: 50vh 50vh` into

```css
grid-template-rows: 1fr 1fr;
height: 100vh;
```

then the image will be too tall in latest Chromium. Because the 1st column will be `100vw - 600px`, then the image will be horizontally stretched to fill that space, and then the aspect-ratio will make the intrinsic row grow too much.

Sure, you have `height: 100%`, but when sizing the rows, the percentage is indefinite, so it behaves as `auto`.

A more general solution, as I posted in https://github.com/w3c/csswg-drafts/issues/6257#issuecomment-848941703, is

```css
img {
  width: 0;
  height: 0;
  min-width: 100%;
  min-height: 100%;
}
```

This way the intrinsic contribution is 0, but when sizing for real, it will fill the desired space.

**Comment by bfgeek** [comment]
Specifically for the `aspect-ratio: none` case consider the following example:

```
<!DOCTYPE html>
<div style="display: inline-grid; border: solid;">
  <svg style="background: green;"></svg>
</div>
```
https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=9397
(FF seems to have a bug with above).

Here the `<svg>` element has no aspect-ratio (set the width/height at the other dimension will stay the same size).
This is how:
```
<!DOCTYPE html>
<div style="display: inline-grid; border: solid;">
  <img src=image style="aspect-ratio: none;"></svg>
</div>
```
.... would behave.

The element would still take up space by default - using its natural-size, but have no aspect-ratio. This might be desired! But isn't exactly what the use case is describing unless I'm mis-understanding.

**Comment by mirisuzanne** [comment]
My goal was to remove the aspect-ratio of the image for the purpose of sizing/layout, but still have the natural ratio available when using `object-fit`. It sounds to me like that specific case _should_ work simply with `contain:size`, and could be filed as an implementation bug? Once that's fixed, the syntax is straight forward:

```css
img {
  contain: size;
  object-fit: cover;
}
```

Ian's distinction here makes sense to me. Even though my first instinct was to remove the aspect-ratio of the image, I am actually attempting to remove its size contribution - while keeping the natural aspect ratio intact. As far as I'm concerned, this can be filed as a bug, and closed as no-change.
